lsp/nls/src/background.rs

```diff
@@ -0,0 +1,347 @@
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    path::PathBuf,
+    process::Child,
+    time::{Duration, Instant},
+};
+
+use crossbeam::channel::{Receiver, Select, Sender};
+use ipc_channel::ipc::{IpcOneShotServer, IpcReceiver, IpcSender};
+use log::warn;
+use lsp_types::Url;
+use nickel_lang_core::{
+    cache::SourcePath,
+    eval::{cache::CacheImpl, VirtualMachine},
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{diagnostic::SerializableDiagnostic, files::uri_to_path, world::World};
+
+const EVAL_TIMEOUT: Duration = Duration::from_secs(1);
+
+#[derive(Debug, Serialize, Deserialize)]
+enum Command {
+    UpdateFile { uri: Url, text: String },
+    EvalFile { uri: Url },
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Diagnostics {
+    pub path: PathBuf,
+    pub diagnostics: Vec<SerializableDiagnostic>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Response {
+    Diagnostics(Diagnostics),
+    /// The background worker sends back one of these when it's about to start
+    /// an eval job. That way, if it becomes unresponsive we know which file is
+    /// the culprit.
+    Starting {
+        uri: Url,
+    },
+}
+
+pub struct BackgroundJobs {
+    receiver: Receiver<Diagnostics>,
+    sender: Sender<Command>,
+}
+
+// The entry point of the background worker. If it fails to bootstrap the connection,
+// panic immediately (it's a subprocess anyway).
+pub fn worker_main(main_server: String) {
+    let oneshot_tx = IpcSender::connect(main_server).unwrap();
+    let (cmd_tx, cmd_rx) = ipc_channel::ipc::channel().unwrap();
+    let (response_tx, response_rx) = ipc_channel::ipc::channel().unwrap();
+    oneshot_tx.send((cmd_tx, response_rx)).unwrap();
+
+    worker(cmd_rx, response_tx);
+}
+
+fn drain_ready<T: for<'de> Deserialize<'de> + Serialize>(rx: &IpcReceiver<T>, buf: &mut Vec<T>) {
+    while let Ok(x) = rx.try_recv() {
+        buf.push(x);
+    }
+}
+
+// Returning an Option, just to let us use `?` when channels disconnect.
+fn worker(cmd_rx: IpcReceiver<Command>, response_tx: IpcSender<Response>) -> Option<()> {
+    let mut evals = VecDeque::new();
+    let mut cmds = Vec::new();
+    let mut world = World::default();
+
+    drain_ready(&cmd_rx, &mut cmds);
```

Comment:
Is this one really necessary? I guess if we remove it, the first for loop will just not run at all, and it'll have the same effect at the next loop iteration.